#!/bin/bash
#PBS -N illuminaseq
#PBS -P xl04
#PBS -q normal
#PBS -l walltime=12:00:00
#PBS -l mem=192GB
#PBS -l ncpus=48
#PBS -l storage=gdata/xl04+gdata/if89
#PBS -l wd
#PBS -M kirat.alreja@anu.edu.au

module load bwa

# Define the reference genome file path
reference="/g/data/xl04/ka6418/species/Tiliqua_Rugosa/rTilRug0.5/assembly/rTilRug0.5.asm.hp1.fasta"

# Define the input file containing the list of input files
input_list="/g/data/xl04/ka6418/ausarg/experimental_scripts/illuminaseq/file_list.txt"

# Read input file list into an array
mapfile -t query_fastqs < "$input_list"

# Loop through all input files
for query_fastq in "${query_fastqs[@]}"; do
    # Extract the sample name from the file name
    sample=$(basename "${query_fastq%_R1.clean.fq.gz}")

    # Check if the input file is a paired-end read file
    if [[ $query_fastq == *_R1.clean.fq.gz ]]; then
        # Align the paired-end reads using BWA MEM
        bwa mem -t $PBS_NCPUS "$reference" "$query_fastq" "${query_fastq/_R1/_R2}" > "${sample}_paired.sam"
    # Check if the input file is a singleton read file
    elif [[ $query_fastq == *_singletons.fq.gz ]]; then
        # Align the singleton reads using BWA MEM
        bwa mem -t $PBS_NCPUS "$reference" "$query_fastq" > "${sample}_singleton.sam"
    fi
done
