#!/bin/bash

#PBS -N pac-ont-multiple
#PBS -P xl04
#PBS -q normal
#PBS -l walltime=24:00:00
#PBS -l mem=192GB
#PBS -l ncpus=48
#PBS -l storage=gdata/xl04+gdata/if89
#PBS -l wd
#PBS -l jobfs=400GB
#PBS -M kirat.alreja@anu.edu.au

module load samtools
module load vcftools 
module load bcftools

INPUT_BAM=/g/data/xl04/ka6418/sequence_alignment/bam_outputs/tiliqua_rugosa/pacbio/merged/assembly.haplotype1/350719_assembly.haplotype1_merged.bam
REFERENCE=/g/data/xl04/ka6418/species/Tiliqua_Rugosa/rTilRug0.5/assembly/rTilRug0.5.asm.hp1.fasta
OUTPUT_VCF=/g/data/xl04/ka6418/ausarg-data/output.vcf

samtools mpileup -uf $REFERENCE $INPUT_BAM | bcftools call -mv > $OUTPUT_VCF

FILTERED_VCF=/g/data/xl04/ka6418/ausarg-data/filtered.vcf
SUMMARY_STATS=/g/data/xl04/ka6418/ausarg-data/summary_stats.txt

# Filter the VCF file
vcftools --vcf $OUTPUT_VCF --remove-indels --maf 0.05 --max-missing 0.1 --recode --out $FILTERED_VCF

# Calculate summary statistics
vcftools --vcf $FILTERED_VCF.recode.vcf --missing-indv --out $SUMMARY_STATS